crates/sp-lightclient/src/lib.rs

```diff
@@ -159,8 +156,29 @@ pub trait HeaderImporter<Header: HeaderT> {
         )
         .map_err(|_| ImportError::InvalidSolution)?;
 
-        // TODO(ved): calculate cumulative block weight
-        // TODO(ved): check for existing forks and prune fork headers
+        let block_weight =
+            calculate_block_weight(&global_randomness.global_randomness, &pre_digest);
+        let total_weight = parent_header.total_weight + block_weight;
+
+        // last best header should ideally be parent header. if not check for forks and prune accordingly
+        let last_best_header = Self::Storage::best_header();
+        if last_best_header.header.hash() != parent_header.header.hash() {
+            let last_best_weight = last_best_header.total_weight;
+            if !match total_weight.cmp(&last_best_weight) {
+                // current weight is greater than last best. pick this header
+                Ordering::Greater => true,
+                // if weights are equal, pick the longest chain
+                Ordering::Equal => header.number() > last_best_header.header.number(),
+                // we already are on the best chain
+                Ordering::Less => false,
+            } {
+                return Err(ImportError::NonCanonicalHeader);
+            }
+
+            // re-org: prune all the descendants of the parent header and them import this header as best
+            Self::Storage::prune_descendants_of(parent_header.header.hash());
```

Comment:
What if we need to re-org on lower parent block after this happened?

Reply:
If the light client sees a header H1 whose parent is an older parent of the current parent of the header H2 we are importing, and if the fork choice rules suggest that the  H1 is to be taken as the best block, then we prune all the descendants of the H1 parent and import H1 as the best

Reply:
I re-read it and I fail to see how it is possible for reorgs to happen back and forth. If you move along one fork you can't really switch to another later unless the first block after the point of divergence is of a higher weight than the block from fork that is currently known as canonical (code doesn't really care what is the height difference between proposed header and `last_best_header`.

Reply:
i see you point. Let me fix that